prime-router/src/main/kotlin/azure/observability/event/ReportEventService.kt

```diff
@@ -0,0 +1,498 @@
+    ITEM_ROUTED,
+}
+
+data class ReportStreamReportEvent(
+    @JsonUnwrapped
+    val reportEventData: ReportEventData,
+    @JsonUnwrapped
+    val params: Map<ReportStreamEventProperties, Any>,
+) : AzureCustomEvent
+
+data class ReportStreamItemEvent(
+    @JsonUnwrapped
+    val reportEventData: ReportEventData,
+    @JsonUnwrapped
+    val itemEventData: ItemEventData,
+    @JsonUnwrapped
+    val params: Map<ReportStreamEventProperties, Any>,
+) : AzureCustomEvent
+
+abstract class AbstractReportStreamEventBuilder<T : AzureCustomEvent>(
+    protected val reportEventService: ReportEventService,
+    val azureEventService: AzureEventService,
+    private val theName: ReportStreamEventName,
+    private val childReportId: UUID,
+    private val childBodyUrl: String,
+    private val theTopic: Topic,
+    private val pipelineStepName: TaskAction,
+) : Logging {
+
+    constructor(
+        reportEventService: ReportEventService,
+        azureEventService: AzureEventService,
+        theName: ReportStreamEventName,
+        report: ReportFile,
+        pipelineStepName: TaskAction,
+    ) : this(
+        reportEventService,
+        azureEventService,
+        theName,
+        report.reportId,
+        report.bodyUrl,
+        report.schemaTopic,
+        pipelineStepName
+    )
+
+    constructor(
+        reportEventService: ReportEventService,
+        azureEventService: AzureEventService,
+        theName: ReportStreamEventName,
+        report: Report,
+        pipelineStepName: TaskAction,
+    ) : this(
+        reportEventService,
+        azureEventService,
+        theName,
+        report.id,
+        report.bodyURL,
+        report.schema.topic,
+        pipelineStepName
+    )
+    var theParams: Map<ReportStreamEventProperties, Any> = emptyMap()
+    var theParentReportId: UUID? = null
+
+    companion object {
+
+        fun <Event, T : AbstractReportStreamEventBuilder<Event>> buildEvent(
+            reportEventService: ReportEventService,
+            azureEventService: AzureEventService,
+            eventBuilderClass: Class<T>,
+            initializer: T.() -> Unit,
+        ): T {
+            val constructor =
+                eventBuilderClass.getDeclaredConstructor(ReportEventService::class.java, AzureEventService::class.java)
+            return constructor.newInstance(reportEventService, azureEventService).apply(initializer)
+        }
+    }
+
+    fun parentReportId(parentReportId: UUID?) {
+        theParentReportId = parentReportId
+    }
+
+    fun params(params: Map<ReportStreamEventProperties, Any>) {
+        theParams = params
+    }
+
+    abstract fun buildEvent(): T
+
+    fun getReportEventData(): ReportEventData {
+        return reportEventService.getReportEventData(
+            childReportId,
+            childBodyUrl,
+            theParentReportId,
+            pipelineStepName,
+            theTopic
+        )
+    }
+
+    fun sendToAzure(): AbstractReportStreamEventBuilder<T> {
```

Comment:
This function and `logEvent` feels like it breaks the principle of the `builder` pattern.

I don't think our in-progress of being built event should be used until it is finished.

I think it would make more sense to live in the ReportEventService.

Reply:
Happy to pair on this, but my thought is that it makes it easier for the caller to consume and prevents people from calling those functions without explicitly going through the builder.

Reply:
Discussed in a huddle, going to leave the approach as is for now.